pthread/pthread.c

```diff
@@ -457,24 +480,37 @@ int pthread_equal(pthread_t t1, pthread_t t2)
 }
 
 
-void pthread_exit(void *value_ptr)
+static __attribute__((noreturn)) void pthread_do_exit(pthread_ctx *ctx, void *value_ptr, int cleanup)
 {
-	pthread_ctx *ctx = (pthread_ctx *)pthread_self();
-	_pthread_key_cleanup_t *cleanup;
-
 	if (ctx != NULL) {
-		mutexLock(pthread_common.pthread_list_lock);
-		_pthread_do_cleanup(ctx);
+		if (cleanup != 0) {
+			mutexLock(pthread_common.pthread_list_lock);
+			_pthread_do_cleanup(ctx);
+			mutexUnlock(pthread_common.pthread_list_lock);
+		}
+
+		pthread_key_cleanup(ctx);
 		ctx->retval = value_ptr;
-		cleanup = pthread_key_cleanup(ctx);
 		_pthread_ctx_put(ctx);
```

Comment:
We don't have `pthread_list_lock` here - use `pthread_ctx_put`? Or execute `_pthread_do_cleanup` after `pthread_key_cleanup` and then do `_pthread_ctx_put` or `pthread_ctx_put` accordingly to avoid lock/unlock twice.

Reply:
put should not be called here, mixup during rebase